runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -382,24 +385,16 @@ public String toString() {
         return "LogData[" + getGlobalAddress() + "]";
     }
 
-    /**
-     * Verify that max payload is enforced for the specified limit.
-     *
-     * @param limit Max write limit
-     * @return the serialized size of the payload
-     */
-    public int checkMaxWriteSize(int limit) {
-        Preconditions.checkState(serializedCache != null, "checkMaxWriteSize requires serialized form");
-
-        int payloadSize = getSizeEstimate();
-        if (log.isTraceEnabled()) {
-            log.trace("checkMaxWriteSize: payload size is {} bytes.", payloadSize);
+    private void checkMaxWriteSizeIfRequired(Optional<Integer> limit, ByteBuf buf) {
+        if (!limit.isPresent()) {
+            return;
         }
 
-        if (payloadSize > limit) {
-            throw new WriteSizeException(payloadSize, limit);
-        }
+        int payloadSize = buf.readableBytes();
```

Comment:
This seems correct for the existing code paths, but I think the computation would be more robust against future changes if the difference between the old and new writerIndex value is computed explicitly.